add AddressIndex changes needed for bitcore block explorer

LarryRuane opened the pull request:
Addresses #3708, note there are no tests yet (will come later with RPC interfaces).

LarryRuane commented:
The first (in time) commit is for #3832 and will disappear after #3832 is merged (I will do a rebase and force-push).

LarryRuane commented on src/init.cpp:
```diff
@@ -1444,8 +1445,18 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
+        if (nBlockTreeDBCache > (1 << 21) && !GetBoolArg("-txindex", false)) {
+            nBlockTreeDBCache = (1 << 21); // block tree db cache shouldn't be larger than 2 MiB
+        }
+    }
```
source for above: https://github.com/bitpay/bitcoin/commit/c91d78b578a8700a45be936cb5bb0931df8f4b87#diff-c865a8939105e6350a50af02766291b7R1233 except I added a requirement that the `-addressindex` option requires the `-txindex` option, which I'm almost certain I came to understand is true (but I can't provide a reference, sorry). Note also that I'm not adding db compression or maxopenfiles changes yet, because I'm not yet convinced they are needed, and also to keep this PR reasonably small. We can revisit later.

bitcartel replied:
Is there anything in the bitpay upstream code which confirms the requirement for `-txindex`?

LarryRuane replied:
See https://github.com/bitpay/bitcore-node/blob/master/docs/upgrade.md (find in page `txindex`), but you probably noticed that the test I added at https://github.com/zcash/zcash/pull/3837/files#diff-c865a8939105e6350a50af02766291b7R1449 isn't in upstream (either hackworks or bitpay). I added it because I thought it would be an improvement to require `-txindex`, but should I remove that test so that we match upstream? That change would be a separate commit (too late to force-push it).

LarryRuane commented on src/txdb.h:
```diff
@@ -17,8 +17,19 @@
+typedef std::pair<CAddressIndexKey, CAmount> CAddressAmount;
+
+typedef std::vector<CAddressUnspent> CAddressUnspentVec;
+typedef std::vector<CAddressAmount> CAddressAmountVec;
```
I added the above 4 typedefs since they were repeated many times. Suggestions for better names would be appreciated!

LarryRuane commented on src/txdb.h:
```diff
@@ -70,6 +81,14 @@ class CBlockTreeDB : public CDBWrapper
+    bool EraseAddressIndex(const std::vector<CAddressAmount> &vect);
+    bool ReadAddressIndex(uint160 addressHash, int type,
+                          std::vector<CAddressAmount> &addressIndex,
+                          int start = 0, int end = 0);
```
Oops, not a big problem, but I just noticed I can use my new typedefs in some of the function prototypes above, to simplify them, will do that in a later force-push.

LarryRuane commented:
For now, please look only at the Feb 13 commit, not the Feb 9 one (that's covered by PR #3832).

LarryRuane commented on src/txdb.cpp:
```diff
@@ -328,23 +328,23 @@ bool CBlockTreeDB::ReadAddressUnspentIndex(uint160 addressHash, int type, CAddre
 }
 
-bool CBlockTreeDB::WriteAddressIndex(const std::vector<CAddressAmount> &vect) {
+bool CBlockTreeDB::WriteAddressIndex(const CAddressIndexAmountVec &vect) {
```
No functional change; I had added some `typedef`s in the previous commit but didn't update all the places that could use the new typedef (here's one of them).

mdr0id commented on src/init.cpp:
```diff
@@ -1444,8 +1445,18 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
+        // enable 3/4 of the cache if addressindex is enabled
+        nBlockTreeDBCache = nTotalCache * 3 / 4;
+
+    } else {
```
We can simplify this with just a single `else if` block, but minor nit

bitcartel replied:
To reduce the diff size, you could leave the two lines removed as they originally were and not have the else { } block.  If addressindex is set, nBlockTreeDBCache will get set.

LarryRuane replied:
Just for future reference, nothing is too much of a nit for me; please always provide them. I love these kinds of improvements.

Eirik0 commented:
I believe that this needs a rebase to remove https://github.com/zcash/zcash/pull/3837/commits/f00020a7ed075232895917e06c17072effc9c856.

Eirik0 commented:
One question I had on this one was why is this code coming from the bitpay/bitcoin form and not bitcoin/bitcoin itself? Is there some reason why bitcoin does not want to include this code?

bitcartel commented:
@Eirik0 This third-party code is not essential to the functioning of Bitcoin, so there is little reason for it to be incorporated into bitcoin master.  Given that there are lots of third-party products based on custom changes to Bitcoin, how does one decide which are blessed to live in master?  I think it would be preferable for the explorer code to be merged into a branch of Zcash which gets rebased on every release, as already done in a separate repo here: https://github.com/zcash-hackworks/zcash-patched-for-explorer

LarryRuane commented:
Thanks, bitcartel, that makes sense and is good to know. @Eirik0, an additional point that str4d mentioned to me yesterday is that there are something like 20 block explorer implementations for bitcoin, while there are only a couple for us, Insight (bitpay) being the most important. To have reasonable performance, a block explorer needs to store additional redundant state (redundant in that it can be derived from the blockchain). Many of the bitcoin block explorers work without needing any changes to `bitcoind` by maintaining separate leveldb tables to hold this state. Insight took what might be considered a shortcut and modified `bitcoind` (and `zcashd`) to maintain this extra state on its behalf, which was easier because the client is already maintaining tables (indices) anyway for other reasons.

bitcartel commented on src/main.cpp:
```diff
@@ -2248,16 +2252,38 @@ DisconnectResult DisconnectBlock(const CBlock& block, const CBlockIndex* pindex,
     for (int i = block.vtx.size() - 1; i >= 0; i--) {
         const CTransaction &tx = block.vtx[i];
-        uint256 hash = tx.GetHash();
+        uint256 const txhash = tx.GetHash();
```
Could still be `hash` given we can immediately see we are calling `tx.GetHash()` so we know the context and the scope of the variable is just within this for loop.

LarryRuane commented:
> I believe that this needs a rebase to remove [f00020a](https://github.com/zcash/zcash/commit/f00020a7ed075232895917e06c17072effc9c856).

@Eirik0, that commit is now gone. It appeared as part of this PR because its branch needed to be based on the branch for PR #3859 (rather than master, as usual), so that PR's commit (the one you referenced) is part of the difference between this PR's branch and master. I did mention this in the second comment above. 

Now that #3859 has been merged, and I've rebased and force-pushed the current branch, the #3859 commit is no longer part of the difference between the current branch and master, so it has disappeared.

I don't know if there's a better way to do this. I did the same thing for the subsequent related PRs (#3849 and #3857), each of which built on the previous PR, and so they include commits that they really shouldn't contain. I tried to point all this out in the comments (so that people don't review the same commits more than once). But maybe there's a better way? I could "release" the PRs one at a time, and not release the next one until the current one has been merged; that would fix this problem. But it seemed better to have all four of them (now we're down to 3) available at the same time, this problem notwithstanding.

bitcartel commented on src/script/script.cpp:
```diff
@@ -227,3 +238,29 @@ bool CScript::IsPushOnly() const
+    } else if (this->IsPayToPublicKeyHash()) {
+        start = 3;
+    } else {
+        assert(false);
```
So it looks like a nonstandard output script could trigger the assertion and termiante the app.  So maybe this method only gets called once we're already sure the script is valid, per the comment.  If yes, then if you already have the type (1 or 2) from calling Type(), why not pass that value in and check against it, instead of calling IsPayToXXXHash() again?

Eirik0 replied:
I agree with the removal of the assertion, but I prefer to just call `this->IsPayToPublicKeyHash()` again rather than to pass in the type. See my other comment related to this.

LarryRuane commented on src/main.cpp:
```diff
@@ -2248,16 +2252,38 @@ DisconnectResult DisconnectBlock(const CBlock& block, const CBlockIndex* pindex,
-        uint256 hash = tx.GetHash();
+        uint256 const txhash = tx.GetHash();
+
+        if (fUpdateAddressIndex) {
```
Instead of introducing a new keyword (or phrase) to search for, I'll make a change so that everything is guarded by `fAddressIndex` (so searching for that will find everything).

LarryRuane commented on src/main.cpp:
```diff
@@ -2248,16 +2252,38 @@ DisconnectResult DisconnectBlock(const CBlock& block, const CBlockIndex* pindex,
+                    CAddressUnspentValue()));
+            }
+        }
 
```
Again, after the next commit, the explorer-related code will be clearly identified by `fAddressIndex`.

LarryRuane commented on src/main.cpp:
```diff
@@ -2288,6 +2314,26 @@ DisconnectResult DisconnectBlock(const CBlock& block, const CBlockIndex* pindex,
                 if (!ApplyTxInUndo(undo, view, out))
                     fClean = false;
+
+                if (fUpdateAddressIndex) {
```
I don't want to do this, because then, to be consistent, I'll have to add similar comments in many places. I'm thinking this PR's comments can serve as second-level documentation. Let me know if you disagree strongly.

bitcartel replied:
@LarryRuane I think we need a comment of some sort in the code, to at least link to the PR itself otherwise users won't know where to look unless they do a git blame and search commit history.  One of the issues about relying on Github is that conversations on PRs can be hidden from view (see pic) and no guarantee that the repo stays up.
![image](https://user-images.githubusercontent.com/194979/54235324-3a164300-44ce-11e9-94c1-53205593791d.png)

daira replied:
I'm actually inclined to not include comments like this, since the code could easily become cluttered with them. We don't usually include the origin of particular pieces of code.

bitcartel replied:
The comments can be removed in the future.  For now, I think they're quite useful in helping to clearly identify explorer related code, especially as the changes are being integrated/assembled over multiple PRs.

bitcartel commented on src/main.cpp:
```diff
@@ -2940,7 +2940,7 @@ bool static DisconnectTip(CValidationState &state, bool fBare = false) {
     {
         CCoinsViewCache view(pcoinsTip);
-        if (DisconnectBlock(block, state, pindexDelete, view, fAddressIndex) != DISCONNECT_OK)
+        if (DisconnectBlock(block, state, pindexDelete, view, true) != DISCONNECT_OK)
```
Why did this change to always being true, whereas before it might have been false?

LarryRuane replied:
See my comment below where I explain this; it's really just so that `fAddressIndex` can appear in the code right at the place where it has an effect (to make it easier for readers).

bitcartel commented on src/main.cpp:
```diff
@@ -2630,19 +2630,19 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
-                    continue;
-                uint160 const addrHash = uint160(out.scriptPubKey.AddressHash());
+                if (scriptType > 0) {
+                    uint160 const addrHash = uint160(out.scriptPubKey.AddressHash(scriptType));
```
Missing check?  What happens when uint160 is called with an empty vector?  AddressHash() has been updated in the commit where it can return an empty vector.  If empty, the script type is invalid for the purposes of the explorer, and we should 'continue' to next iteration of loop.  (This comment applies everywhere AddressHash(scriptType) is used.)

LarryRuane replied:
I think it's okay because `AddressHash()` only returns an empty vector if `scriptType` is zero, see https://github.com/zcash/zcash/pull/3837/files#diff-f7ca24fb80ddba0f291cb66344ca6fcbR256 but please double-check that for me.

Yes, I used to have exactly that, `if (scriptType == 0) continue` but changed it to this because later, when we add `spentindex` (#3863), we are going to want to do some stuff in the body of this loop even if `scriptType` is zero. So writing it this way makes adding that change easier.

bitcartel replied:
@LarryRuane So `AddressHash(scriptType)` is only invoked when `scriptType > 0`, but if `scriptType >= 3`, an empty vector is returned. Given that addressIndex only supports scriptType 1 or 2, maybe the conditional `if (scriptType > 0)` should be changed to reflect this e.g. `if (scriptType == 1 || scriptType == 2) { ... }`

LarryRuane replied:
@bitcartel I think this code is correct as it is. We may recognize more script types in the future, say `scriptType == 3`, and this code won't have to change, and it shouldn't have to change. The convention is: If `Type()` returns non-zero, that means `AddressHash()`, when passed that return value, returns a legitimate address (not an empty vector). As long as the two methods, `Type()` and `AddressHash`, are kept in sync (and they're implemented right next to each other in `src/script/script.cpp`), then it will be okay. I think it will be pretty clear to future developers that those functions need to stay in sync. (But if you disagree, I could add a comment.)

bitcartel commented on src/main.cpp:
```diff
@@ -2258,32 +2258,32 @@ static DisconnectResult DisconnectBlock(const CBlock& block, CValidationState& s
+        uint256 const hash = tx.GetHash();
 
-        if (fUpdateAddressIndex) {
+        if (fAddressIndex && fUpdateAddressIndex) {
```
Is this a fix?  It seems previously only the parameter was checked but not the global for fAddressIndex, which might have been false.

LarryRuane replied:
No, not a fix; previously the `fUpdateAddressIndex` argument to this function was passed in as `fAddressIndex` by one caller and `false` by the other caller. But one of your earlier comments pointed out that it would be nice if readers could search for `fAddressIndex` and easily find all the affected code. So I made this change, really only so `fAddressIndex` appears here. I also changed the first caller to pass `true` instead of `fAddressIndex`, so the functionality is the same.

LarryRuane commented on src/addressindex.h:
```diff
@@ -10,6 +10,8 @@
 #include "amount.h"
 #include "script/script.h"
 
+// fAddressIndex (insight_explorer) (entire file)
```
Okay on nit1. On nit2, I think it would be better to keep file locations as they are so that merging future upstream changes is easier.

LarryRuane commented on src/init.cpp:
```diff
@@ -340,7 +340,6 @@ std::string HelpMessage(HelpMessageMode mode)
 
     string strUsage = HelpMessageGroup(_("Options:"));
     strUsage += HelpMessageOpt("-?", _("This help message"));
-    strUsage += HelpMessageOpt("-addressindex", strprintf(_("bitcore: Maintain a full address index, used to query for the balance, txids and unspent outputs for addresses (default: %u)"), false)); // fAddressIndex (insight_explorer)
```
I didn't add a help message for `-insightexplorer` because none of the other experimental options are documented in this part of the help (one-line summaries).

LarryRuane commented on src/main.cpp:
```diff
@@ -4230,9 +4231,14 @@ bool static LoadBlockIndexDB()
+    if (fInsightExplorer) {
+        // upstream insight exposes individual index options; we have
+        // chosen to use a single (experimental) flag to control all of them
+        fAddressIndex = true;
```
In later PRs, `fSpentIndex` and `fTimestampIndex` will also get set to `true` here.

Eirik0 commented on src/main.cpp:
```diff
@@ -67,6 +67,8 @@ bool fExperimentalMode = false;
 bool fReindex = false;
 bool fTxIndex = false;
+bool fInsightExplorer = false;
+bool fAddressIndex = false;     // same value as fInsightExplorer
```
I like the change to use `-indexexplorer` as a single flag. I think that we should either use boolean variable `fAddressIndex` or replace it with `fInsightExplorer`

Eirik0 commented on src/main.cpp:
```diff
@@ -2248,16 +2252,38 @@ DisconnectResult DisconnectBlock(const CBlock& block, const CBlockIndex* pindex,
         return DISCONNECT_FAILED;
     }
+    CAddressAmountVec addressIndex;
+    CAddressUnspentVec addressUnspentIndex;
```
I am ambivalent about the typedefs. I think that in this case I prefer just having the raw types, but I could go either way.

LarryRuane replied:
@Eirik0, I took that halfway approach we discussed yesterday. I removed the `*Vec` types because they don't do much (save only the `vector<...>`) but I did keep the other typedefs (that are not in upstream) that begin with `pair<` because they do appear in many places, so seems like a useful abstraction.

Eirik0 commented on src/script/script.cpp:
```diff
@@ -227,3 +238,33 @@ bool CScript::IsPushOnly() const
     return true;
 }
+
+int CScript::Type() const
```
I'm not the biggest fan of the `Type()` method, but I do think it is an improvement on the code that this is coming from. I considered what if this returned an enum, or what if the function return a pair of type address hash (to avoid the other method), or even to put in some polymorphic structure to encapsulate all of this. That being said said, I'm not sure that any of these suggestions are worth it given that it is fine as is.

LarryRuane replied:
@Eirik0 After further thought, I decided to leave this as-is; it would be disruptive to change it to an enum or make other kinds of changes.

LarryRuane commented:
@bitcartel and @Eirik0, can you please take one more look. I squashed down to one commit, and I've incorporated all important requests and suggestions. I also tried to make it possible for a future reader of this code to search for `insightexplorer` and find everything that's related (and will continue that convention for the subsequent PRs). Also, I added URL references to the important new blocks of code (and if you have time, please also check those). Thanks!

LarryRuane commented:
@Eirik0 had a good suggestion, cherry-pick the upstream commit that adds the new `test/script_P2PKH_tests.cpp` instead of squashing it. No code change, only history. (Thanks, Eirik.)

Eirik0 commented on src/main.h:
```diff
@@ -270,6 +270,218 @@ struct CNodeStateStats {
+    size_t index;
+
+    size_t GetSerializeSize(int nType, int nVersion) const {
+        return 57;
```
This is the same as in bitpay, so I think it's fine.

Eirik0 commented on src/script/script.cpp:
```diff
@@ -227,3 +238,29 @@ bool CScript::IsPushOnly() const
+        assert(false);
+    }
+
+    vector<unsigned char> hashBytes(this->begin()+start, this->begin()+start+20);
```
You could change this method to return a `uint160`. It looks like everywhere it is called we do something like `uint160 const addrHash = uint160(out.scriptPubKey.AddressHash(scriptType));` so this could be inlined in to this method.

Eirik0 replied:
Also, `CScript::AddressHash(int scriptType)` could be modified so that we do not pass in the scriptType. Instead of 
```
    switch (scriptType) {
    case 1: // P2PKH
    ...
```
we could do 
```
    if (this->IsPayToPublicKeyHash())
        return 1;
    ...
```
(The same as in `CScript::Type()`)

We end up redoing some work, but it means that we can't possible pass in the wrong script type.

LarryRuane replied:
Yes, I like both of your suggestions, will push another commit.

Eirik0 commented on src/txdb.h:
```diff
@@ -17,6 +17,18 @@
+struct CAddressIndexIteratorHeightKey;
+
+typedef std::pair<CAddressUnspentKey, CAddressUnspentValue> CAddressUnspent;
+typedef std::pair<CAddressIndexKey, CAmount> CAddressIndexAmount;
```
How about `CAddressUnspentEntry` and `CAddressIndexEntry`.

Eirik0 replied:
I think the crucial thing to communicate is the structs represent pairs which I think entry does. I personally wouldn't have included `Db`, but I think that's fine.

LarryRuane commented:
@Eirik0, I'm not sure I explained that choice about the flags well enough... Yes, I don't like that redundancy either, but Simon's idea for keeping the three as separate flags (in this PR, only `fAddressIndex`, later also `fSpentIndex` and `fTimestampIndex`), instead of `fInsightExplorer` being the only flag, is to make comparison with, and merging from, upstream easier; and also so that readers know which of the three indices are related to which chunks of code. One final reason is that it's not super-unlikely that in the future we may want to separate out the flags again because some use cases turn up that require some of these indices but not others.

As for the URL comments, we considered removing them but Simon and I agree that these parts of the code are likely to be active in the short-term, so the comments will be useful. When things settle down and this code isn't likely to change, we can remove the comments and thereby reduce the clutter (which is always a good goal).

bitcartel commented:
I'm testing this with options enabled and a reindex on mainnet datadir, so could take a while.

bitcartel commented:
@zkbot try

bitcartel commented:
@zkbot r+

LarryRuane commented:
@zkbot try

bitcartel commented:
@zkbot retry

bitcartel commented:
@zkbot retry

bitcartel commented:
@zkbot retry
